File: Source/GLNPianoView.m
//
//  GLNPianoView.h
//  Piano Keyboard View
//
//  Created by Gary Newby on 11/03/2013.
//  Copyright (c) 2013 Gary Newby. All rights reserved.
//

#import "GLNPianoView.h"
#import <QuartzCore/QuartzCore.h>


@interface GLNPianoView ()

@property (nonatomic, strong) UIImage *whiteDoImg;
@property (nonatomic, strong) UIImage *whiteUpImg;
@property (nonatomic, strong) UIImage *blackDoImg;
@property (nonatomic, strong) UIImage *blackUpImg;
@property (nonatomic, assign) CGFloat blackHeight;
@property (nonatomic, assign) CGFloat blackWidth;
@property (nonatomic, assign) CGFloat lastWidth;
@property (nonatomic, assign) CGFloat cradius;
@property (nonatomic, assign) NSInteger numOfWhiteKeys;
@property (nonatomic, strong) NSMutableArray *keyDown;
@property (nonatomic, strong) NSMutableArray *keyRects;
@property (nonatomic, strong) NSMutableArray *sublayers;
@property (nonatomic, strong) NSMutableSet *currentTouches;

@end



@implementation GLNPianoView

- (void)prepareForInterfaceBuilder
{
    if (_totalNumKeys < 12) {
        self.totalNumKeys = 12;
    }
}

- (void) commonInit
{
    for (int i = 0; i < self.sublayers.count; i++) {
        [self.sublayers[i] removeFromSuperlayer];
    }
    
    self.keyDown = [NSMutableArray array];
    self.keyRects = [NSMutableArray array];
    self.sublayers = [NSMutableArray array];
    for (int i = 0; i < self.totalNumKeys; i++) {
        [self.keyDown addObject:[NSNull null]];
        [self.keyRects addObject:[NSValue valueWithCGRect:CGRectZero]];
        [self.sublayers addObject:[NSNull null]];
    }
    
    self.blackHeight = 0.585;
    self.blackWidth = 0.63;
    self.cradius = self.blackWidth * 8.0;
    self.numOfWhiteKeys = 0;
    for (int i = 1; i < self.totalNumKeys; i++) {
        self.keyDown[i] = @NO;
        if ([self isWhiteKey:i]) {
            self.numOfWhiteKeys++;
        }
    }
    
    self.currentTouches = [NSMutableSet set];
    self.multipleTouchEnabled = YES;
    self.layer.masksToBounds = YES;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    [self commonInit];
    
    CGRect rect = self.bounds;
    CGFloat whiteKeyHeight = rect.size.height;
    CGFloat whiteKeyWidth  = [self whiteKeywidth:rect];
    CGFloat blackKeyHeight = rect.size.height * self.blackHeight;
    CGFloat blackKeyWidth  = whiteKeyWidth * self.blackWidth;
    CGFloat blackKeyOffset = blackKeyWidth / 2.0;
    
    self.blackUpImg = [self blackKeyImage:CGSizeMake(blackKeyWidth, blackKeyHeight) keyDown:NO];
    self.blackDoImg = [self blackKeyImage:CGSizeMake(blackKeyWidth, blackKeyHeight) keyDown:YES];
    self.whiteUpImg = [self whiteKeyImage:CGSizeMake(21, 21) keyDown:NO];
    self.whiteDoImg = [self whiteKeyImage:CGSizeMake(21, 21) keyDown:YES];
    
    CGFloat runningX = 0;
    NSInteger yval = 0;
    
    // White Keys
    for (int i = 0; i < self.totalNumKeys; i++) {
        if ([self isWhiteKey:i]) {
            NSInteger newX = (NSInteger)(runningX + 0.5);
            NSInteger newW = (NSInteger)((runningX + whiteKeyWidth + 0.5) - newX);
            CGRect keyRect = CGRectMake(newX, yval, newW, whiteKeyHeight - 1);
            self.sublayers[i] = [self createCALayer:[UIColor whiteColor] rect:keyRect white:YES];
            [self.layer addSublayer:self.sublayers[i]];
            runningX += whiteKeyWidth;
        }
    }
    runningX = 0.0f;
    
    // Black Keys
    for (int i = 0; i < self.totalNumKeys; i++) {
        if ([self isWhiteKey:i]) {
            runningX += whiteKeyWidth;
        } else {
            CGRect keyRect = CGRectMake((runningX - blackKeyOffset), yval, blackKeyWidth, blackKeyHeight);
            self.sublayers[i] = [self createCALayer:[UIColor blackColor] rect:keyRect white:NO];
            [self.layer addSublayer:self.sublayers[i]];
        }
    }
}

- (float)whiteKeywidth:(CGRect)rect
{
    return (rect.size.width / self.numOfWhiteKeys);
}

- (CALayer *)createCALayer:(UIColor*)color rect:(CGRect)rect white:(BOOL)white
{
    rect = CGRectMake(CGRectGetMinX(rect), CGRectGetMinY(rect) - (self.cradius*2.0), CGRectGetWidth(rect), CGRectGetHeight(rect) + (self.cradius*2.0));
    
    CALayer *layer = [CALayer layer];
    CGImageRef blackUp = self.blackUpImg.CGImage;
    CGImageRef whiteUp = self.whiteUpImg.CGImage;
    
    NSUInteger x = 0;
    layer.contentsCenter = CGRectMake(0.5, 0.5, 0.02, 0.02);
    
    if (white) {
        x = 1.0;
        layer.contents = (__bridge id)whiteUp;
    }
    
    layer.frame = CGRectInset(rect, x, 0);
    layer.opaque = YES;
    layer.backgroundColor = color.CGColor;
    layer.cornerRadius = self.cradius;
    
    // Disable implict animations for specified keys
    layer.actions = @{@"onOrderIn":[NSNull null], @"onOrderIn":[NSNull null], @"onOrderOut":[NSNull null],
                      @"self.sublayers":[NSNull null], @"contents":[NSNull null], @"bounds":[NSNull null], @"position":[NSNull null]};
    if (!white) {
        // Black
        layer.contents       = (__bridge id)blackUp;
        layer.masksToBounds  = YES;
    }
    
    return layer;
}

-(void)keyDown:(NSInteger)keyNum
{
    CALayer *layer = ((CALayer *)self.sublayers[keyNum]);
    
    if ([self isWhiteKey:keyNum]) {
        CGImageRef whiteDo = self.whiteDoImg.CGImage;
        layer.contents = (__bridge id)whiteDo;
        layer.contentsCenter = CGRectMake(0.5, 0.5, 0.1, 0.1);
    } else {
        CGImageRef blackDo = self.blackDoImg.CGImage;
        layer.contents = (__bridge id)blackDo;
    }
}

-(void)keyUp:(NSInteger)keyNum
{
    CALayer *layer = ((CALayer *)self.sublayers[keyNum]);
    
    if ([self isWhiteKey:keyNum]) {
        CGImageRef whiteUp = self.whiteUpImg.CGImage;
        layer.contents = (__bridge id)whiteUp;
        layer.contentsCenter = CGRectMake(0.5, 0.5, 0.02, 0.02);
    } else {
        CGImageRef blackUp = self.blackUpImg.CGImage;
        layer.contents = (__bridge id)blackUp;
    }
}

- (BOOL)isWhiteKey:(NSInteger)key
{
    NSUInteger k = key %12;
    return (k == 0 || k == 2 || k == 4 || k == 5 || k == 7 || k == 9 || k == 11);
}

-(NSInteger)getKeyboardKey:(CGPoint)point
{
    NSInteger keyNum = NSNotFound;
    for (int i = 0; i < self.totalNumKeys; i++) {
        if (CGRectContainsPoint([self.keyRects[i] CGRectValue], point)) {
            keyNum = i;
            if (![self isWhiteKey:i]) {
                break;
            }
        }
    }
    return keyNum;
}

#pragma mark - Update

-(void)updateKeyRects
{
    if (self.lastWidth == self.bounds.size.width) {
        return;
    }
    self.lastWidth = self.bounds.size.width;
    
   	CGRect rect            = self.frame;
    CGFloat whiteKeyHeight = self.bounds.size.height;
    CGFloat blackKeyHeight = whiteKeyHeight * self.blackHeight ;
    CGFloat whiteKeyWidth  = [self whiteKeywidth:rect];
    CGFloat blackKeyWidth  = whiteKeyWidth * self.blackWidth;
    CGFloat leftKeyBound   = whiteKeyWidth - (blackKeyWidth / 2.0f);
    
    NSInteger lastWhiteKey = 0;
    self.keyRects[0] = [NSValue valueWithCGRect:CGRectMake(0, 0, whiteKeyWidth, whiteKeyHeight)];
    
    for (int i = 1; i < self.totalNumKeys; i++) {
        if (![self isWhiteKey:i]) {
            self.keyRects[i] = [NSValue valueWithCGRect:CGRectMake((lastWhiteKey * whiteKeyWidth) + leftKeyBound, 0, blackKeyWidth, blackKeyHeight)];
        } else {
            lastWhiteKey++;
            self.keyRects[i] = [NSValue valueWithCGRect:CGRectMake(lastWhiteKey * whiteKeyWidth, 0, whiteKeyWidth, whiteKeyHeight)];
        }
    }
}

- (void)updateKeyStates
{
    [self updateKeyRects];
    
    NSArray* touches = [self.currentTouches allObjects];
    NSUInteger count = [touches count];
    NSMutableArray *currentKeyState = [NSMutableArray array];
    
    for (int i = 0; i < self.totalNumKeys; i++) {
        currentKeyState[i] = @NO;
    }
    
    for (int i = 0; i < count; i++) {
        UITouch* touch = [touches objectAtIndex:i];
        CGPoint point = [touch locationInView:self];
        NSInteger index = [self getKeyboardKey:point];
        if (index != NSNotFound) {
            currentKeyState[index] = @YES;
        }
    }
    
    for (int i = 0; i < self.totalNumKeys; i++) {
        if (self.keyDown[i] != currentKeyState[i]) {
            BOOL keyDownState = [currentKeyState[i] boolValue];
            self.keyDown[i] = @(keyDownState);
            if (keyDownState) {
                [self.delegate pianoKeyDown:i];
                [self keyDown:i];
            } else {
                [self.delegate pianoKeyUp:i];
                [self keyUp:i];
            }
        }
    }
    
    [self setNeedsDisplay];
}

#pragma mark - Touch Handling Code

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    for (UITouch* touch in touches) {
        [self.currentTouches addObject:touch];
    }
    [self updateKeyStates];
}

- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event
{
    [self updateKeyStates];
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    for (UITouch* touch in touches) {
        [self.currentTouches removeObject:touch];
    }
    [self updateKeyStates];
}

- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event
{
    for (UITouch* touch in touches) {
        [self.currentTouches removeObject:touch];
    }
    [self updateKeyStates];
}

#pragma mark - Image generation

- (UIImage *)blackKeyImage:(CGSize)size keyDown:(BOOL)keyDown
{
    CGFloat scale = [UIScreen mainScreen].scale;
    size.width *= scale;
    size.height *= scale;
    UIGraphicsBeginImageContext(size);
    
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    UIColor* fillColor = [UIColor colorWithRed: 0.379 green: 0.379 blue: 0.379 alpha: 1];
    UIColor* strokeColor = [UIColor colorWithRed: 0 green: 0 blue: 0 alpha: 0.951];
    
    NSArray* gradient1Colors = [NSArray arrayWithObjects: (id)strokeColor.CGColor, (id)fillColor.CGColor, nil];
    CGFloat gradient1Locations[] = {0.11, 1};
    CGGradientRef gradient1 = CGGradientCreateWithColors(colorSpace, (__bridge CFArrayRef)gradient1Colors, gradient1Locations);
    
    CGRect frame = CGRectMake(0, 0, size.width, size.height);
    
    UIBezierPath* rectanglePath = [UIBezierPath bezierPathWithRect: CGRectMake(CGRectGetMinX(frame), CGRectGetMinY(frame), size.width, size.height)];
    [strokeColor setFill];
    [rectanglePath fill];
    
    CGFloat border = size.width*0.15;
    CGFloat width = size.width*0.7;
    CGFloat topRectHeight = size.height*0.86;
    CGFloat bottomRectOffset = size.height*0.875;
    CGFloat bottomRectHeight = size.height*0.125;
    if (keyDown) {
        topRectHeight = size.height*0.91;
        bottomRectOffset = size.height*0.925;
    }
    
    CGRect roundedRectangleRect = CGRectMake(CGRectGetMinX(frame) +  border, CGRectGetMinY(frame), width, topRectHeight);
    UIBezierPath* roundedRectanglePath = [UIBezierPath bezierPathWithRoundedRect: roundedRectangleRect cornerRadius: self.cradius];
    CGContextSaveGState(context);
    [roundedRectanglePath addClip];
    CGContextDrawLinearGradient(context, gradient1, CGPointMake(CGRectGetMidX(roundedRectangleRect), CGRectGetMinY(roundedRectangleRect)),
                                CGPointMake(CGRectGetMidX(roundedRectangleRect), CGRectGetMaxY(roundedRectangleRect)), 0);
    CGContextRestoreGState(context);
    
    CGRect roundedRectangle2Rect = CGRectMake(CGRectGetMinX(frame) +  border, CGRectGetMinY(frame) + bottomRectOffset, width, bottomRectHeight);
    UIBezierPath* roundedRectangle2Path = [UIBezierPath bezierPathWithRoundedRect: roundedRectangle2Rect cornerRadius: self.cradius];
    CGContextSaveGState(context);
    [roundedRectangle2Path addClip];
    CGContextDrawLinearGradient(context, gradient1, CGPointMake(CGRectGetMidX(roundedRectangle2Rect), CGRectGetMaxY(roundedRectangle2Rect)),
                                CGPointMake(CGRectGetMidX(roundedRectangle2Rect), CGRectGetMinY(roundedRectangle2Rect)), 0);
    CGContextRestoreGState(context);
    
    CGGradientRelease(gradient1);
    CGColorSpaceRelease(colorSpace);
    
    UIImage * image = UIGraphicsGetImageFromCurrentImageContext();
    return image;
}

- (UIImage *)whiteKeyImage:(CGSize)size keyDown:(BOOL)keyDown
{
    CGFloat scale = [UIScreen mainScreen].scale;
    size.width *= scale;
    size.height *= scale;
    UIGraphicsBeginImageContext(size);
    
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    UIColor* fillColor = [UIColor colorWithRed: 0.1 green: 0.1 blue: 0.1 alpha: 0.20];
    UIColor* strokeColor = [UIColor colorWithRed: 0.9 green: 0.9 blue: 0.9 alpha: 0.0];
    if (keyDown) {
        fillColor = [UIColor colorWithRed: 0.1 green: 0.1 blue: 0.1 alpha: 0.50];
        strokeColor = [UIColor colorWithRed: 0.9 green: 0.9 blue: 0.9 alpha: 0.0];
    }
    
    NSArray* gradient1Colors = [NSArray arrayWithObjects: (id)strokeColor.CGColor, (id)fillColor.CGColor, nil];
    CGFloat gradient1Locations[] = {0.1, 1};
    CGGradientRef gradient1 = CGGradientCreateWithColors(colorSpace, (__bridge CFArrayRef)gradient1Colors, gradient1Locations);
    
    UIBezierPath* rectanglePath = [UIBezierPath bezierPathWithRect: CGRectMake(0, 0, size.width, size.height)];
    CGContextSaveGState(context);
    [rectanglePath addClip];
    CGContextDrawRadialGradient(context, gradient1, CGPointMake(size.width/2.0, size.height/2.0), size.height*0.01,
                                CGPointMake(size.width/2.0, size.height/2.0), size.height*0.6,
                                kCGGradientDrawsBeforeStartLocation | kCGGradientDrawsAfterEndLocation);
    CGContextRestoreGState(context);
    
    CGGradientRelease(gradient1);
    CGColorSpaceRelease(colorSpace);
    
    UIImage * image = UIGraphicsGetImageFromCurrentImageContext();
    return image;
}

@end

File: Source/GLNPianoView.swift
//
//  GLNPianoView.swift
//  GLNPianoView
//
//  Created by Gary Newby on 16/05/2016.
//  Copyright © 2016 Gary Newby. All rights reserved.
//

import UIKit

@objc public protocol GLNPianoViewDelegate: class {
    func pianoKeyUp(_ keyNumber: Int)
    func pianoKeyDown(_ keyNumber: Int)
}

@IBDesignable public class GLNPianoView: UIView {
    
    @IBInspectable var showNotes: Bool = true
    @objc public weak var delegate: GLNPianoViewDelegate?
    private var keysArray: [GLNPianoKey?] = []
    private var currentTouches = NSMutableSet(capacity: Int(maxNumberOfKeys))
    private static let minNumberOfKeys = 12
    private static let maxNumberOfKeys = 61
    private var _octave = 60
    private var _numberOfKeys = 24
    private var whiteKeyCount = 0
    private var _blackKeyHeight: CGFloat = 0.60
    private var _blackKeyWidth: CGFloat = 0.80
    private var keyCornerRadius: CGFloat = 0
    private var labels: [String?] = Array.init(repeating: nil, count: 128)
    
    @IBInspectable public var blackKeyHeight: CGFloat {
        get {
            return _blackKeyHeight
        }
        set {
            let value = newValue.clamp(min: 0, max: 10)
            _blackKeyHeight = (value.rounded() + 5) * 0.05
        }
    }
    
    @IBInspectable public var blackKeyWidth: CGFloat {
        get {
            return _blackKeyWidth
        }
        set {
            let value = newValue.clamp(min: 0, max: 8)
            _blackKeyWidth = (value.rounded() + 10) * 0.05
        }
    }

    @IBInspectable public var numberOfKeys: Int {
        get {
            return _numberOfKeys
        }
        set {
            _numberOfKeys = newValue.clamp(min: GLNPianoView.minNumberOfKeys, max: GLNPianoView.maxNumberOfKeys)
            initKeys()
            setNeedsLayout()
        }
    }

    public var octave: Int {
        get {
            return _octave
        }
        set {
            _octave = newValue
            initKeys()
            setNeedsLayout()
        }
    }
    
    @objc public func toggleShowNotes() {
        showNotes.toggle()
        setNeedsLayout()
    }
    
    @objc public func aKeyIsDown() -> Bool {
        var downKeyCount = 0
        for key in keysArray {
            if let k = key, k.isDown {
                downKeyCount += 1
            }
        }
        return (downKeyCount > 0)
    }

    // MARK: - Init

    public override init(frame: CGRect) {
        super.init(frame: frame)
        initKeys()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        initKeys()
    }

    private func initKeys() {
        keyCornerRadius = _blackKeyWidth * 8.0
        whiteKeyCount = 0
        currentTouches = NSMutableSet()
        keysArray = [GLNPianoKey?](repeating: nil, count: Int(_numberOfKeys + 1))
        for index in 1 ..< _numberOfKeys + 1 {
            if index.isWhiteKey() {
                whiteKeyCount += 1
            }
        }
        isMultipleTouchEnabled = true
        layer.masksToBounds = true
        if let subLayers = layer.sublayers {
            for layer in subLayers {
                layer.removeFromSuperlayer()
            }
        }

        let rect: CGRect = bounds
        let whiteKeyHeight = rect.size.height
        let whiteKeyWidth = whiteKeyWidthForRect(rect)
        let blackKeyHeight = rect.size.height * _blackKeyHeight
        let blackKeyWidth = whiteKeyWidth * _blackKeyWidth
        let blackKeyOffset = blackKeyWidth / 2.0

        // White Keys
        var xPosition: CGFloat = 0
        for index in 0 ..< _numberOfKeys {
            if index.isWhiteKey() {
                let newX = (xPosition + 0.5)
                let newW = ((xPosition + whiteKeyWidth + 0.5) - newX)
                let keyRect = CGRect(x: newX, y: 0, width: newW, height: whiteKeyHeight - 1)
                let noteNumber = index + octave
                let key = GLNPianoKey(color: UIColor.white, rect: keyRect, type: .white, cornerRadius: keyCornerRadius, showNotes: showNotes, noteNumber: noteNumber, label: labels[noteNumber])
                keysArray[index] = key
                layer.addSublayer(key.imageLayer)
                xPosition += whiteKeyWidth
            }
        }
        // Black Keys
        xPosition = 0.0
        for index in 0 ..< _numberOfKeys {
            if index.isWhiteKey() {
                xPosition += whiteKeyWidth
            } else {
                let keyRect = CGRect(x: (xPosition - blackKeyOffset), y: 0, width: blackKeyWidth, height: blackKeyHeight)
                let noteNumber = index + octave
                let key = GLNPianoKey(color: UIColor.black, rect: keyRect, type: .black, cornerRadius: keyCornerRadius, showNotes: showNotes, noteNumber: noteNumber, label: labels[noteNumber], blackKeyWidth: blackKeyWidth, blackKeyHeight: blackKeyHeight)
                keysArray[index] = key
                layer.addSublayer(key.imageLayer)
            }
        }
    }
    
    private func whiteKeyWidthForRect(_ rect: CGRect) -> CGFloat {
        return (rect.size.width / CGFloat(whiteKeyCount))
    }
    
    private func updateKeys() {
        let touches = currentTouches.allObjects as Array
        var keyIsDownAt = [Bool](repeating: false, count: _numberOfKeys)
        
        for touchIndex in 0 ..< touches.count {
            let touch = touches[touchIndex]
            let point = (touch as AnyObject).location(in: self)
            let index = getKeyContaining(point)
            if index != NSNotFound {
                keyIsDownAt[index] = true
            }
        }
        
        for index in 0 ..< _numberOfKeys {
            if keysArray[index]?.isDown != keyIsDownAt[index] {
                if keyIsDownAt[index] {
                    delegate?.pianoKeyDown(index)
                    keysArray[index]?.setImage(keyNum: index, isDown: true)
                } else {
                    delegate?.pianoKeyUp(index)
                    keysArray[index]?.setImage(keyNum: index, isDown: false)
                }
                keysArray[index]?.isDown = keyIsDownAt[index]
            }
        }
        setNeedsDisplay()
    }
    
    private func getKeyContaining(_ point: CGPoint) -> Int {
        var keyNum = NSNotFound
        for index in 0 ..< _numberOfKeys {
            if let frame = keysArray[index]?.imageLayer.frame, frame.contains(point) {
                keyNum = index
                if !index.isWhiteKey() {
                    break
                }
            }
        }
        return keyNum
    }

    // MARK: - Highlighting

    public func highlightKeys(noteNames: [String], color: UIColor, play: Bool = false) {
        reset(didPlay: play)
        for note in noteNames {
            let noteNumber = GLNNote.midiNumber(for: note)
            highlight(noteNumber: noteNumber, color: color, play: play)
        }
    }

    public func highlightKeys(noteNumbers: [Int], color: UIColor, play: Bool = false) {
        reset(didPlay: play)
        for noteNumber in noteNumbers {
            highlight(noteNumber: noteNumber, color: color, play: play)
        }
    }

    public func highlightKey(noteName: String, color: UIColor, play: Bool = false) {
        reset(didPlay: play)
        let noteNumber = GLNNote.midiNumber(for: noteName)
        highlight(noteNumber: noteNumber, color: color, play: play)
    }

    public func highlightKey(noteNumber: Int, color: UIColor, play: Bool = false, resets: Bool = true) {
        reset(didPlay: play)
        highlight(noteNumber: noteNumber, color: color, play: play, resets: resets)
    }

    private func highlight(noteNumber: Int, color: UIColor, play: Bool, resets: Bool = true) {
        for key in keysArray {
            if let key = key, key.noteNumber == noteNumber  {
                key.highlightLayer.backgroundColor = color.cgColor
                key.resetsHighLight = resets
                if play {
                    delegate?.pianoKeyDown(key.noteNumber - octave)
                }
            }
        }
    }

    public func reset(didPlay: Bool) {
        for key in keysArray {
            if let key = key  {
                if key.resetsHighLight {
                    key.highlightLayer.backgroundColor = UIColor.clear.cgColor
                }
                if didPlay {
                    delegate?.pianoKeyUp(key.noteNumber - octave)
                }
            }
        }
    }

    // MARK: - Labels

    public func setLabel(for midiNumber: Int, text: String) {
        guard midiNumber < labels.count else {
            return
        }
        labels[midiNumber] = text
    }

    // MARK: - Touches
    
    public override func touchesBegan(_ touches: Set<UITouch>, with _: UIEvent?) {
        for touch in touches {
            currentTouches.add(touch)
        }
        updateKeys()
    }
    
    public override func touchesMoved(_ touches: Set<UITouch>, with _: UIEvent?) {
        for touch in touches {
            currentTouches.add(touch)
        }
        updateKeys()
    }
    
    public override func touchesEnded(_ touches: Set<UITouch>, with _: UIEvent?) {
        for touch in touches {
            currentTouches.remove(touch)
        }
        updateKeys()
    }
    
    public override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        for touch in touches {
            currentTouches.remove(touch)
        }
        updateKeys()
    }
}
